pgrx/src/nullable.rs

```diff
@@ -0,0 +1,523 @@
+// by Postgres.
+pub enum AnyNullLayout<'a> {
+    /// Bitmap of where null slots are in this container.
+    /// For example, 00001001 would represent:
+    /// \[value, value, value, value, null, value value, null\]
+    /// However, the underlying data buffer would be:
+    /// \[value, value, value, value, value, value\]
+    /// because of the skip behavior
+    Bitmap(BitSliceNulls<'a>),
+    // TODO: Find implementation details on this
+    BoolSlice(BoolSliceNulls<'a>),
+    /// Bool map, simply an array of booleans telling you
+    /// No nulls
+    Strict(StrictNulls),
+}
+
+impl<'a> NullLayout<usize> for AnyNullLayout<'a> {
+    fn len(&self) -> usize {
+        match self {
+            AnyNullLayout::Bitmap(bits) => bits.len(),
+            AnyNullLayout::BoolSlice(bools) => bools.len(),
+            AnyNullLayout::Strict(strict) => strict.len(),
+        }
+    }
+
+    fn has_nulls(&self) -> bool {
+        match self {
+            AnyNullLayout::Bitmap(bits) => bits.has_nulls(),
+            AnyNullLayout::BoolSlice(bools) => bools.has_nulls(),
+            AnyNullLayout::Strict(_strict) => false,
+        }
+    }
+
+    fn is_valid(&self, idx: usize) -> Option<bool> {
+        match self {
+            AnyNullLayout::Bitmap(bits) => bits.is_valid(idx),
+            AnyNullLayout::BoolSlice(bools) => bools.is_valid(idx),
+            AnyNullLayout::Strict(strict) => strict.is_valid(idx),
+        }
+    }
+
+    fn is_null(&self, idx: usize) -> Option<bool> {
+        match self {
+            AnyNullLayout::Bitmap(bits) => bits.is_null(idx),
+            AnyNullLayout::BoolSlice(bools) => bools.is_null(idx),
+            AnyNullLayout::Strict(strict) => strict.is_null(idx),
+        }
+    }
+
+    fn count_nulls(&self) -> usize {
+        match self {
+            AnyNullLayout::Bitmap(bits) => bits.count_nulls(),
+            AnyNullLayout::BoolSlice(bools) => bools.count_nulls(),
+            AnyNullLayout::Strict(strict) => strict.count_nulls(),
+        }
+    }
+}
+
+/// Iterates over a layout of null values, returning true for values that are null.
+pub struct NullsIter<'a, Idx, Layout: NullLayout<Idx>>
+where
+    Idx: PartialEq + PartialOrd,
+{
+    layout: &'a Layout,
+    current: Idx,
+}
+
+impl<'mcx, Layout: NullLayout<usize>> Iterator for NullsIter<'mcx, usize, Layout> {
+    type Item = Nullable<()>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // is_null() should handle bounds checking internally
+        let result_value = self.layout.is_null(self.current).map(|b| match b {
+            false => Nullable::Valid(()),
+            true => Nullable::Null,
+        });
+        self.current += 1;
+        result_value
+    }
+}
+
+/// Iterator for Strict containers,
+/// which is to say, containers which cannot contain nulls
+pub struct NoNullsIter {
+    /// Will be set to container.len() and so a 0 value means we've reached the end.
+    remaining: usize,
+}
```

Comment:
An alternative possibility is to simply make this iterator infinite.

Reply:
I could see that but I'm wondering if there's any context in which the null-layout iter will be used to do bounds-checking, given that the null slice is already used to do that outside of the context of iterators.

Reply:
The main reason that is done the way it is was because otherwise we were storing the len like 20 times (once in RawArray, once in NullKind, once in Array itself, and another place...) so I was trying to reduce the number of places that was stored. You can comfortably assume anything that isn't _inherent_ to the type doesn't need to be tracked by these implementations, and other code can simply always do
```rust
nulls.take(len).zip(values).map(|(null, datum)| match null {
        /* etc... */
    }
)
```

Reply:
Oh that's a good point, I might try to cut out unnecessary copies of the length tag tomorrow.

Reply:
Done, `MaybeStrictNulls` now does not have a length tag for the strict case. Array's bounds'-checking is done by accessing Array. See 39f5ad9